chain/pruned_block_dispatcher_test.go

```diff
@@ -0,0 +1,593 @@
+
+		expBlock, ok := h.blocks[blockHash]
+		require.Equal(h.t, expBlock, block)
+
+		// Decrement how many clients queried the same block. Once we
+		// have none left, remove it from the map.
+		h.blocksQueried[blockHash]--
+		if h.blocksQueried[blockHash] == 0 {
+			delete(h.blocksQueried, blockHash)
+		}
+
+	case <-time.After(5 * time.Second):
+		select {
+		case err := <-errChan:
+			h.t.Fatalf("received unexpected error send: %v", err)
+		default:
+		}
+		h.t.Fatal("expected reply from peer")
+	}
+
+	// If we should expect a nil error to be sent by the internal
+	// workManager to signal completion of the request, wait for it now.
+	if expectCompletionSignal {
+		select {
+		case err := <-errChan:
+			require.NoError(h.t, err)
+		case <-time.After(5 * time.Second):
+			h.t.Fatal("expected nil err to signal completion")
+		}
+	}
+}
+
+// assertNoPeerDialed asserts that the PrunedBlockDispatcher hasn't established
+// a new peer connection.
+func (h *networkBlockTestHarness) assertNoPeerDialed() {
+	h.t.Helper()
+
+	select {
+	case peer := <-h.dialedPeer:
+		h.t.Fatalf("unexpected connection established with peer %v", peer)
+	case <-time.After(2 * time.Second):
+	}
+}
+
+// assertNoReply asserts that the peer hasn't replied to a query.
+func (h *networkBlockTestHarness) assertNoReply(
+	blockChan <-chan *wire.MsgBlock, errChan <-chan error) {
+
+	h.t.Helper()
+
+	select {
+	case block := <-blockChan:
+		h.t.Fatalf("received unexpected block %v", block.BlockHash())
+	case err := <-errChan:
+		h.t.Fatalf("received unexpected error send: %v", err)
+	case <-time.After(2 * time.Second):
+	}
+}
+
+// TestPrunedBlockDispatcherQuerySameBlock tests that client requests for the
+// same block result in only fetching the block once while pending.
+func TestPrunedBlockDispatcherQuerySameBlock(t *testing.T) {
+	t.Parallel()
+
+	const numBlocks = 1
+	const numPeers = 5
+	const numRequests = numBlocks * numPeers
+
+	h := newNetworkBlockTestHarness(t, numBlocks, numPeers, numPeers)
+	h.start()
+	defer h.stop()
+
+	// Queue all the block requests one by one.
+	blockChans := make([]<-chan *wire.MsgBlock, 0, numRequests)
+	errChans := make([]<-chan error, 0, numRequests)
+	for i := 0; i < numRequests; i++ {
+		blockChan, errChan := h.query(h.hashes)
+		blockChans = append(blockChans, blockChan)
+		errChans = append(errChans, errChan)
+	}
+
+	// We should only see one query.
+	h.assertPeerQueried()
+	for i := 0; i < numRequests; i++ {
+		h.assertPeerReplied(blockChans[i], errChans[i], i == 0)
+	}
+}
+
+// TestPrunedBlockDispatcherMultipleGetData tests that a client requesting blocks
+// that span across multiple queries works as intended.
+func TestPrunedBlockDispatcherMultipleGetData(t *testing.T) {
+	t.Parallel()
+
+	const numBlocks = wire.MaxInvPerMsg + 1
+
+	h := newNetworkBlockTestHarness(t, numBlocks, 1, 1)
+	h.start()
+	defer h.stop()
+
+	// Request all blocks. We extend the timeout of the query as this will
+	// take longer than the default 30s timeout.
```

Comment:
why does it take this long in the test environment?

Reply:
It's run in parallel along with other tests and 50k channel sends occur (which is slower when running with race enabled). There's also only one query peer to simplify the test scenario. The test when ran by itself without race takes 50s on a MacBook Pro.